storage/client/internal/curl_wrappers.cc

```diff
@@ -31,19 +32,50 @@ extern "C" std::size_t WriteCallback(void* contents, std::size_t size,
   return size * nmemb;
 }
 
+extern "C" std::size_t HeaderCallback(char* contents, std::size_t size,
+                                      std::size_t nmemb, void* dest) {
+  auto* headers = reinterpret_cast<storage::internal::CurlHeaders*>(dest);
+  headers->Append(contents, size * nmemb);
+  return size * nmemb;
+}
+
 }  // anonymous namespace
 
 namespace storage {
 inline namespace STORAGE_CLIENT_NS {
 namespace internal {
-void CurlBuffer::CaptureOutputOf(CURL* curl) {
+void CurlBuffer::Attach(CURL* curl) {
   curl_easy_setopt(curl, CURLOPT_WRITEDATA, this);
   curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
 }
 
 void CurlBuffer::Append(char* data, std::size_t size) {
   buffer_.append(data, size);
 }
+
+void CurlHeaders::Attach(CURL* curl) {
+  curl_easy_setopt(curl, CURLOPT_HEADERDATA, this);
+  curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, HeaderCallback);
+}
+
+void CurlHeaders::Append(char* data, std::size_t size) {
+  if (size <= 2) {
+    // Empty header (including the \r\n), ignore.
+    return;
+  }
+  if (data[size - 1] != '\n' or data[size - 2] != '\r') {
```

Comment:
I feel like this would be better written in-order rather than reverse-order:

```cpp
if (data[size - 2] != '\r' or data[size - 1] != '\n') {
```

but that's just me.